### ISSUE TITLE
Axis Transition Performance Degradation

### ISSUE THREAD

**Issue by giantismstrikes** [d3/d3#1597]
The performance of axis transitions degrades over time when constantly adding new data. Directly below is an example of this. Below the example of performance degradation is an example of the same data being plotted without a transitioning x axis.

``` html
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>D3: Vertical axis added</title>
    <script type="text/javascript" src="./scripts/d3.v3.min.js"></script>
    <script src="http://ajax.googleapis.com/ajax/libs/jquery/1.3.1/jquery.min.js"></script>
        <style type="text/css">

            body {
    font-family: sans-serif;
    font-size: 9pt;
    line-height: 12pt;
    background: #ffffff;
    color: #555555;
}
.line1 {
    fill: none;
    stroke: steelblue;
    stroke-width: 1px;
}
.line2 {
    fill: none;
    stroke: green;
    stroke-width: 1px;
}
.line3 {
    fill: none;
    stroke: orange;
    stroke-width: 1px;
}
.line4 {
    fill: none;
    stroke: red;
    stroke-width: 1px;
}
.axis path, .axis line {
    fill: none;
    stroke: #555555;
    shape-rendering: crispEdges;
}
        </style>
    </head>
    <body>


    <div id="graph" class="aGraph" style="position:absolute;top:0px;left:0; float:left;"></div>

    <div id="divValue"> </div>


    <script>
         var t = -1;
 var n = 20;
 var duration = 5;

 var data1 = initialise();

 function initialise() {
     var time = -1;
     var arr = [];
     for (var i = 0; i < n; i++) {
         var obj = {
             time: ++time,
             value: Math.floor(Math.random() * (1850 - 400 + 1) + 400)
         };
         arr.push(obj);
     }
     t = time;
     return arr;
 }

 // push a new element on to the given array
 function updateData(a) {
     var obj = {
         time: t,
         value: Math.floor(Math.random() * (1850 - 400 + 1) + 400)
     };
     a.push(obj);

     // make sure that the array does not get too big
     //a = a.slice(t - n);

  //$("#divValue").text(t);
 }

 var margin = {
     top: 10,
     right: 10,
     bottom: 20,
     left: 40
 },
 width = 960 - margin.left - margin.right,
     height = 500 - margin.top - margin.bottom;

 var x = d3.scale.linear()
     .domain([t - n + 1, t])
     .range([0, width]);

 var y = d3.scale.linear()
.domain([d3.min(data1, function(d) { return d.value; }), d3.max(data1, function(d) { return d.value; })])
     .range([400, 0]);

 var line = d3.svg.line()
     .interpolate("basis")
     .x(function (d, i) {
     return x(d.time);
 })
     .y(function (d, i) {
     return y(d.value);
 });

 var svg = d3.select("body").append("svg")
     .attr("width", width + margin.left + margin.right)
     .attr("height", height + margin.top + margin.bottom);

 var g = svg.append("g")
     .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

 // extra svg to clip the graph and x axis as they transition in and out
 var graph = g.append("svg")
     .attr("width", width)
     .attr("height", height + margin.top + margin.bottom);


 var xAxis = d3.svg.axis().scale(x).orient("bottom");

 var axis = graph.append("g")
     .attr("class", "x axis")
     .attr("transform", "translate(0," + 400 + ")")
     .call(x.axis = xAxis);

var yAxisScale = d3.scale.linear()
    .domain([0, 100])
    .range([400, 0]);

var yAxis = d3.svg.axis()
    .scale(yAxisScale)
    .orient("left");

var yAxisGroup = g.append("g")
    .attr("class", "y axis")
    .call(yAxis);

 var path1 = graph.append("g")
     .append("path")
     .data([data1])
     .attr("class", "line1");

 tick();

 function tick() {

     t++;

     // push
     updateData(data1);


     // update the domains
     x.domain([t - n + 2, t]);

     // redraw the lines
     graph.select(".line1").attr("d", line).attr("transform", null);

     // slide the line left
     path1.transition()
         .duration(duration)
         .ease("linear")
         .attr("transform", "translate(" + x(t - n + 1) + ")");



     // slide the x-axis left
     axis.transition()
         .duration(duration)
         .ease("linear")
         .call(xAxis)
         .each("end", tick);

     data1.shift();    

 }

 showOrHide = function (i) {
var selectt = document.getElementsByClassName("messageCheckbox")[i].value;
     if (document.getElementsByClassName("messageCheckbox")[i].checked) {
         document.getElementsByClassName(selectt)[0].style.display = 'none';
     } else {
         document.getElementsByClassName(selectt)[0].style.display = 'block';
     }
 }

        </script>
    </body>
</html>
```

Compare that to running it without an x axis that is updating.

``` html
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>D3: Vertical axis added</title>
    <script type="text/javascript" src="./scripts/d3.v3.min.js"></script>
    <script src="http://ajax.googleapis.com/ajax/libs/jquery/1.3.1/jquery.min.js"></script>
        <style type="text/css">

            body {
    font-family: sans-serif;
    font-size: 9pt;
    line-height: 12pt;
    background: #ffffff;
    color: #555555;
}
.line1 {
    fill: none;
    stroke: steelblue;
    stroke-width: 1px;
}
.line2 {
    fill: none;
    stroke: green;
    stroke-width: 1px;
}
.line3 {
    fill: none;
    stroke: orange;
    stroke-width: 1px;
}
.line4 {
    fill: none;
    stroke: red;
    stroke-width: 1px;
}
.axis path, .axis line {
    fill: none;
    stroke: #555555;
    shape-rendering: crispEdges;
}
        </style>
    </head>
    <body>


    <div id="graph" class="aGraph" style="position:absolute;top:0px;left:0; float:left;"></div>

    <div id="divValue"> </div>


    <script>
         var t = -1;
 var n = 20;
 var duration = 5;

 var data1 = initialise();

 function initialise() {
     var time = -1;
     var arr = [];
     for (var i = 0; i < n; i++) {
         var obj = {
             time: ++time,
             value: Math.floor(Math.random() * (1850 - 400 + 1) + 400)
         };
         arr.push(obj);
     }
     t = time;
     return arr;
 }

 // push a new element on to the given array
 function updateData(a) {
     var obj = {
         time: t,
         value: Math.floor(Math.random() * (1850 - 400 + 1) + 400)
     };
     a.push(obj);

     // make sure that the array does not get too big
     //a = a.slice(t - n);

  $("#divValue").text(t);
 }

 var margin = {
     top: 10,
     right: 10,
     bottom: 20,
     left: 40
 },
 width = 960 - margin.left - margin.right,
     height = 500 - margin.top - margin.bottom;

 var x = d3.scale.linear()
     .domain([t - n + 1, t])
     .range([0, width]);

 var y = d3.scale.linear()
.domain([d3.min(data1, function(d) { return d.value; }), d3.max(data1, function(d) { return d.value; })])
     .range([400, 0]);

 var line = d3.svg.line()
     .interpolate("basis")
     .x(function (d, i) {
     return x(d.time);
 })
     .y(function (d, i) {
     return y(d.value);
 });

 var svg = d3.select("body").append("svg")
     .attr("width", width + margin.left + margin.right)
     .attr("height", height + margin.top + margin.bottom);

 var g = svg.append("g")
     .attr("transform", "translate(" + margin.left + "," + margin.top + ")");

 // extra svg to clip the graph and x axis as they transition in and out
 var graph = g.append("svg")
     .attr("width", width)
     .attr("height", height + margin.top + margin.bottom);

/*
 var xAxis = d3.svg.axis().scale(x).orient("bottom");

 var axis = graph.append("g")
     .attr("class", "x axis")
     .attr("transform", "translate(0," + 400 + ")")
     .call(x.axis = xAxis);
*/
var yAxisScale = d3.scale.linear()
    .domain([0, 100])
    .range([400, 0]);

var yAxis = d3.svg.axis()
    .scale(yAxisScale)
    .orient("left");

var yAxisGroup = g.append("g")
    .attr("class", "y axis")
    .call(yAxis);

/*
 g.append("g")
     .attr("class", "y axis")
     .call(d3.svg.axis().scale(y).orient("left"));
*/

 var path1 = graph.append("g")
     .append("path")
     .data([data1])
     .attr("class", "line1");

 tick();

 function tick() {

     t++;

     // push
     updateData(data1);

// Need to be able to pass in the min/max values 
/*
     y = d3.scale.linear()
    .domain([d3.min(data1, function(d) { return d.value; }), d3.max(data1, function(d) { return d.value; })])
    .range([400, 0]);
*/

     // update the domains
     x.domain([t - n + 2, t]);

     // redraw the lines
     graph.select(".line1").attr("d", line).attr("transform", null);

     // slide the line left
     path1.transition()
         .duration(duration)
         .ease("linear")
         .attr("transform", "translate(" + x(t - n + 1) + ")")
     .each("end", tick);


 /*    path1.exit().transition()
    .duration(duration)
    .ease("linear")
        .attr("transform", "translate(" + x(t - n) + ")")
    .remove();
/*
     // slide the x-axis left
     axis.transition()
         .duration(duration)
         .ease("linear")
         .call(xAxis)
         .each("end", tick);
*/

     data1.shift();    

 }

 showOrHide = function (i) {
var selectt = document.getElementsByClassName("messageCheckbox")[i].value;
     if (document.getElementsByClassName("messageCheckbox")[i].checked) {
         document.getElementsByClassName(selectt)[0].style.display = 'none';
     } else {
         document.getElementsByClassName(selectt)[0].style.display = 'block';
     }
 }

        </script>
    </body>
</html>
```
